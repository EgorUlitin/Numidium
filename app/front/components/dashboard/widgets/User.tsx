import * as React from 'react'

import { Card } from 'antd'

import User from 'model/User'
import Loader from 'components/common/Loader'

import Container from './UserContainer'


export interface Props {
    loading?: boolean
    error?: boolean
    user?: User
}

class UserComponent extends React.PureComponent<Props, {}> {
    render() {
        const { user, loading, error } = this.props

        return (
            <React.Fragment>
                <Card cover={
                    <img alt='avatar' src='https://zos.alipayobjects.com/rmsportal/ODTLcjxAfvqbxHnVXCYX.png' />
                }>
                    <Loader loading={loading || !user} error={error}>
                        {!!user && <Card.Meta title={user && user.email} />}
                    </Loader>
                </Card>
            </React.Fragment>
        )
    }
}

export default Container(UserComponent)
