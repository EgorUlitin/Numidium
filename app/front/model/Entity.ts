interface Entity {
    '@id': string
    '@type': string

    '@context'?: string
}

export default Entity
