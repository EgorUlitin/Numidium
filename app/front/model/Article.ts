import Entity from './Entity'

interface Article extends Entity {
    title: string
    author: string
    year: number
    description: string
    link: string

    libs: any[] // TODO: fix it!
}

export default Article
