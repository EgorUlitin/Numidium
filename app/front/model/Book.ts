import Entity from './Entity'

interface Book extends Entity {
    title: string
    author: string
    year: number
    description: string
    paper: boolean
    shopLink: string
    externalFileLink: string

    libs: any[] // TODO: fix it!
}

export default Book
