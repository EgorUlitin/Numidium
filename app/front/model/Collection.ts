import Entity from './Entity'


interface Collection<T> extends Entity {
    member: T[]
    totalItems: number
}

export default Collection
