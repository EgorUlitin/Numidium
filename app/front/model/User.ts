import Entity from './Entity'

export interface Owner extends Entity {
    id: number
}

export default interface User extends Entity {
    id: number
    email: string
    active: boolean
}
