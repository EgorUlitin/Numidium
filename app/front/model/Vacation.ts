import { Moment } from 'moment'

import Entity from './Entity'
import { Owner } from './User'

interface Vacation extends Entity {
    start: Moment
    end: Moment

    user: Owner
}

export default Vacation
