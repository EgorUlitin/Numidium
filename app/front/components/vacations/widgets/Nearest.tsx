import * as React from 'react'

import { Card, List, Modal } from 'antd'
import * as moment from 'moment'

import Icon, { IconType } from 'components/common/Icon'
import Vacation from 'model/Vacation'

import Container from './NearestContainer'

export interface Props {
    vacations: Vacation[]

    openModal: (id?: number) => void
}

export class Nearest extends React.PureComponent<Props, {}> {

    public render() {
        const { vacations, openModal } = this.props

        return (
            <Card
                title={'Ваши отпуска на ближайщий год'}
                actions={[
                    <span onClick={() => openModal()}>Добавить</span>,
                ]}
            >
                { (vacations.length > 0)
                    ? this.renderTimeline(vacations)
                    : <p>Нет созданных отпусков</p>
                }
            </Card>
        )
    }

    public renderTimeline = (vacations: Vacation[]) =>
        <List
            size={'large'}
            dataSource={vacations}
            renderItem={(v: Vacation) =>
                <List.Item actions={[
                    <Icon onClick={() => this.props.openModal(v.id)} type={IconType.EDIT} />,
                ]}>
                    <List.Item.Meta
                        title={this.getDiffMessage(v)}
                        description={
                            <p>
                                Начало: {v.start.format('DD.MM.YYYY')}<br />
                                Конец: {v.end.format('DD.MM.YYYY')}
                            </p>
                        }
                    />
                </List.Item>
            }
        />

    public getDiffMessage = (v: Vacation) => v.start.diff(moment(), 'days') > 0
        ? `Через ${v.start.diff(moment(), 'days')} дн.`
        : 'Сейчас'
}

export default Container(Nearest)
