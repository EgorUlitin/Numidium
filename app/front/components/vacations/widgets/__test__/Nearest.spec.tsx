import * as React from 'react'
import * as renderer from 'react-test-renderer'

import { Owner } from 'model/User'
import * as moment from 'moment'
import Vacation from 'model/Vacation'

import { Nearest as NearestComponent } from '../Nearest'


test('`Nearest` widget renders correctly without items', () => {
    const tree = renderer
        .create(<NearestComponent vacations={[]} openModal={() => null} />)
        .toJSON()

    expect(tree).toMatchSnapshot()
})

test('`Nearest` widget renders correctly with soon items', () => {
    const owner = {
        '@id': '1',
        '@type': 'TestUser',
        id: 1,
    } as Owner

    const vacations = [
        {
            '@id': '1',
            '@type': 'TestVacation',
            id: 1,
            start: moment('2018-01-01'),
            end: moment('2018-03-02'),
            user: owner,
        } as Vacation,
        {
            '@id': '2',
            '@type': 'TestVacation',
            id: 2,
            start: moment('2019-01-01'),
            end: moment('2019-03-02'),
            user: owner,
        } as Vacation,
    ]

    const tree = renderer
        .create(<NearestComponent vacations={vacations} openModal={() => null} />)
        .toJSON()

    expect(tree).toMatchSnapshot()
})
