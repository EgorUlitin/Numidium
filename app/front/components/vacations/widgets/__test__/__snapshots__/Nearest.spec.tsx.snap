// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`\`Nearest\` widget renders correctly with soon items 1`] = `
<div
  className="ant-card ant-card-bordered"
>
  <div
    className="ant-card-head"
  >
    <div
      className="ant-card-head-wrapper"
    >
      <div
        className="ant-card-head-title"
      >
        Ваши отпуска на ближайщий год
      </div>
    </div>
  </div>
  <div
    className="ant-card-body"
    style={undefined}
  >
    <div
      className="ant-list ant-list-lg ant-list-split"
    >
      <div>
        <div
          className="ant-spin-nested-loading"
          style={null}
        >
          <div
            className="ant-spin-container"
          >
            <div
              className="ant-list-item"
            >
              <div
                className="ant-list-item-meta"
              >
                <div
                  className="ant-list-item-meta-content"
                >
                  <h4
                    className="ant-list-item-meta-title"
                  >
                    Сейчас
                  </h4>
                  <div
                    className="ant-list-item-meta-description"
                  >
                    <p>
                      Начало: 
                      01.01.2018
                      <br />
                      Конец: 
                      02.03.2018
                    </p>
                  </div>
                </div>
              </div>
              <ul
                className="ant-list-item-action"
              >
                <li>
                  <i
                    className="anticon anticon-edit"
                    onClick={[Function]}
                    style={undefined}
                  />
                </li>
              </ul>
            </div>
            <div
              className="ant-list-item"
            >
              <div
                className="ant-list-item-meta"
              >
                <div
                  className="ant-list-item-meta-content"
                >
                  <h4
                    className="ant-list-item-meta-title"
                  >
                    Через 293 дн.
                  </h4>
                  <div
                    className="ant-list-item-meta-description"
                  >
                    <p>
                      Начало: 
                      01.01.2019
                      <br />
                      Конец: 
                      02.03.2019
                    </p>
                  </div>
                </div>
              </div>
              <ul
                className="ant-list-item-action"
              >
                <li>
                  <i
                    className="anticon anticon-edit"
                    onClick={[Function]}
                    style={undefined}
                  />
                </li>
              </ul>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
  <ul
    className="ant-card-actions"
  >
    <li
      style={
        Object {
          "width": "100%",
        }
      }
    >
      <span>
        <span
          onClick={[Function]}
        >
          Добавить
        </span>
      </span>
    </li>
  </ul>
</div>
`;

exports[`\`Nearest\` widget renders correctly without items 1`] = `
<div
  className="ant-card ant-card-bordered"
>
  <div
    className="ant-card-head"
  >
    <div
      className="ant-card-head-wrapper"
    >
      <div
        className="ant-card-head-title"
      >
        Ваши отпуска на ближайщий год
      </div>
    </div>
  </div>
  <div
    className="ant-card-body"
    style={undefined}
  >
    <p>
      Нет созданных отпусков
    </p>
  </div>
  <ul
    className="ant-card-actions"
  >
    <li
      style={
        Object {
          "width": "100%",
        }
      }
    >
      <span>
        <span
          onClick={[Function]}
        >
          Добавить
        </span>
      </span>
    </li>
  </ul>
</div>
`;
